sw/device/tests/flash_ctrl_mem_protection_test.c

```diff
@@ -0,0 +1,212 @@
+                     .ecc_en = kMultiBitBool4True,
+                     .high_endurance_en = kMultiBitBool4False,
+                 },
+         }},
+};
+
+/**
+ * Memory write and readback test
+ * For each MP region, this function check write and read permission
+ * based on MP region properties and priority of each MP region.
+ */
+static void test_mem_access(test_t region) {
+  uint32_t page_size = region.data_region.size;
+  uintptr_t start_epage =
+      (uintptr_t)(START_PAGE_ADDR +
+                  FLASH_CTRL_PARAM_BYTES_PER_PAGE * region.page_start);
+  uint32_t wdata[FLASH_CTRL_PARAM_WORDS_PER_PAGE];
+
+  for (uint32_t i = 0; i < page_size; i++) {
+    for (size_t j = 0; j < ARRAYSIZE(wdata); j++)
+      wdata[j] = region.randomdata[i] + j;
+    if (region.write_permission[i]) {
+      CHECK_STATUS_OK(flash_ctrl_testutils_erase_page(
+          &flash, start_epage,
+          /*partition_id=*/0, kDifFlashCtrlPartitionTypeData));
+      CHECK_STATUS_OK(flash_ctrl_testutils_write(&flash, start_epage,
+                                                 /*partition_id=*/0, wdata,
+                                                 kDifFlashCtrlPartitionTypeData,
+                                                 ARRAYSIZE(wdata)));
+      LOG_INFO("test_mem_access: page %d write complete",
+               region.page_start + i);
+    } else {
+      CHECK_STATUS_NOT_OK(flash_ctrl_testutils_write(
+          &flash, start_epage,
+          /*partition_id=*/0, wdata, kDifFlashCtrlPartitionTypeData,
+          ARRAYSIZE(wdata)));
+      LOG_INFO("test_mem_access: page %d write is not allowed",
+               region.page_start + i);
+    }
+    uint32_t rdata[FLASH_CTRL_PARAM_WORDS_PER_PAGE];
+
+    if (region.read_permission[i]) {
+      CHECK_STATUS_OK(flash_ctrl_testutils_read(
+          &flash, start_epage, /*partition_id=*/0, rdata,
+          kDifFlashCtrlPartitionTypeData, ARRAYSIZE(rdata),
+          /*delay=*/1));
+      CHECK_ARRAYS_EQ(rdata, wdata, ARRAYSIZE(rdata));
+      LOG_INFO("test_mem_access: page %d read check complete",
+               region.page_start + i);
+    } else {
+      CHECK_STATUS_NOT_OK(flash_ctrl_testutils_read(
+          &flash, start_epage, /*partition_id=*/0, rdata,
+          kDifFlashCtrlPartitionTypeData, ARRAYSIZE(rdata),
+          /*delay=*/1));
+      LOG_INFO("test_mem_access: page %d read is not allowed",
+               region.page_start + i);
+    }
+    start_epage += (uintptr_t)FLASH_CTRL_PARAM_BYTES_PER_PAGE;
+  }
+}
+
+bool test_main(void) {
+  LOG_INFO("flash mp test start!");
+  CHECK_DIF_OK(dif_flash_ctrl_init_state(
+      &flash, mmio_region_from_addr(TOP_EARLGREY_FLASH_CTRL_CORE_BASE_ADDR)));
+
+  // Set up default access for data partitions.
+  CHECK_STATUS_OK(flash_ctrl_testutils_default_region_access(
+      &flash, /*rd_en=*/true, /*prog_en=*/true, /*erase_en=*/true,
+      /*scramble_en=*/false, /*ecc_en=*/false, /*high_endurance_en=*/false));
+
+  for (int i = 2; i >= 0; i--) {
```

Comment:
I am curious: should the test have two loops here, one that sets all region properties, and another one that does test_mem_access? This way test_mem_access will be certain to run with properties set as described above.

Reply:
Each loop it programs [table entry](https://github.com/lowRISC/opentitan/blob/2bfcf2a012698672160582861d5f1b38bb720c01/sw/device/tests/flash_ctrl_mem_protection_test.c#L26))
to cover multiple MP regions.

Reply:
I think I understand: each time you create a new region overlapping previous regions starting at a lower page, so it has higher priority. If you agree this is how it works, please change the text around line 34 to reflect this.

Reply:
@matutem , test program higher region first, which has "lower" priority.
Then subsequent programming with lower region will overruled previous config parameter because
lower region number has "higher" priority.